packages/frontend/src/pages/campaigns/index.tsx

```diff
@@ -77,25 +64,72 @@ export const Campaigns = () => {
     const { searchQuery, setSearchQuery } = useSearch();
     const { loading, kpiTokens } = useKPITokens(searchQuery);
 
-    //  show filters
-    const [filtersOpen, setFilterOpen] = useState(false);
+    //  toggle filters
     const toggleFilters = () => setFilterOpen(!filtersOpen);
+    const [filtersOpen, setFilterOpen] = useState(false);
 
     // page settings
     const [pageSize, setPageSize] = useState(12);
     const [ordering, setOrdering] = useState<SelectOption>(ORDERING_OPTIONS[0]);
+
+    // filters
     const [campaignState, setCampaignState] = useState<SelectOption>(
         STATE_OPTIONS[0]
     );
+    const [selectedTemplates, setSelectedTemplates] = useState<Set<number>>(
+        new Set<number>()
+    );
+
+    const [selectedOracles, setSelectedOracles] = useState<Set<number>>(
+        new Set<number>()
+    );
+
+    const handleSelectedTemplatesUpdate = useCallback(
+        (newSelectedTemplates: Set<number>) =>
+            setSelectedTemplates(new Set(newSelectedTemplates)),
+        []
+    );
+
+    const handleSelectedOraclesTemplatesUpdate = useCallback(
+        (newSelectedOracles: Set<number>) =>
+            setSelectedOracles(new Set(newSelectedOracles)),
+        []
+    );
 
     // filter results
-    const filteredTokens = useMemo(() => {
-        return filterKPITokens(
-            Object.values(kpiTokens),
+    const filteredKPITokensByState = useMemo(() => {
+        const tokens = Object.values(kpiTokens);
+
+        return filterStateKPITokens(
+            tokens,
             campaignState.value as unknown as CampaignState
         );
     }, [kpiTokens, campaignState]);
 
+    const hasOracleTemplate = useMemo(
+        () => (token: KPIToken) =>
+            oraclesTemplateIds(token).some((oracleId) => {
+                return selectedOracles.has(Number(oracleId));
+            }),
+        [selectedOracles]
+    );
+    const hasTemplate = useMemo(
+        () => (token: KPIToken) => selectedTemplates.has(token.template.id),
+        [selectedTemplates]
+    );
+
+    const oraclesTemplateIds = (token: KPIToken) =>
+        token.oracles.map((oracle) => oracle.template.id);
+
+    const filteredTokens = useMemo(
+        () =>
+            filteredKPITokensByState.filter(
+                (token) => hasTemplate(token) || hasOracleTemplate(token)
```

Comment:
Same as above. The `hasTemplate` and `hasOracleTemplate` functions are only used here, so it doesn't make sense to have them. Let's move all of their logic in this `useMemo`. We also make React's job easier as we slim down the list of dependencies on the `useMemo` hook and overall reduce the `useMemo` usage.

Reply:
fixed: df180fc782eaa6d8a035701179412777fc890cba